Physical_Engine/Query/IsWithinPileCap.cs

```diff
@@ -0,0 +1,112 @@
+            "if their profile does not intersect with the pad foundation boundary.")]
+        [Input("padFoundation", "The PadFoundation that defines the extents for the Piles to be located.")]
+        [Input("piles", "The Piles to check if they are located within the extents of the PadFoundation.")]
+        [Input("tolerance", "Distance tolerance to be used in the method. Piles are deemed to be on the edge of the PadFoundation if they are within this distance from the curve.", typeof(Length))]
+        [Output("withinPileCap", "True if all Piles are contained within the boundary and depth of the PileCap. False if one or more Piles are not.")]
+        public static bool IsWithinPileCap(this PadFoundation padFoundation, List<Pile> piles, double tolerance = Tolerance.MicroDistance)
+        {
+            if (padFoundation.IsNull())
+                return false;
+            else if (piles == null)
+            {
+                Base.Compute.RecordError("The list of Piles are null, the method will return false.");
+                return false;
+            }
+            else if (piles.Any(x => x.IsNull()))
+                return false;
+
+            // Get the top of the piles
+            List<Point> tops = piles.Select(p => p.Location.IControlPoints().OrderBy(pt => pt.Z).Reverse().First()).ToList();
+
+            // Project all points to top surface of the pad
+            Plane padTop = new Plane() { Normal = padFoundation.Location.Normal(), Origin = padFoundation.Location.Centroid() };
+            List<Point> pTops = tops.Select(x => x.Project(padTop)).ToList(); //project points only used for IsContaining
+
+            // Get the thickness of the PadFoundation to compare against Z coordinates of the pile tops
+            double maxZ = padFoundation.Location.Centroid().Z;
+            double minZ = maxZ - padFoundation.Construction.IThickness();
+
+            ICurve topOutline = padFoundation.Location.ExternalBoundary;
+
+            // Check if all piles are within the curve of the pad
+            if(topOutline.IIsContaining(pTops,true,tolerance))
+            {
+                // Check if the piles are within the depth of the pad
+                if(tops.Any(pt => pt.Z > maxZ || pt.Z < minZ))
+                {
+                    Base.Compute.RecordError("One or more the Piles tops is not located within the depth of the PileCap.");
+                    return false;
+                }
+                else
+                {
+                    // Check if the profile pushes the pile outside the curve of the pad
+                    for(int i = 0; i < pTops.Count; i++)
+                    {
+                        ConstantFramingProperty property = (ConstantFramingProperty)piles[i].Property;
+                        ICurve profile = property.Profile.Edges[0].ITranslate(Engine.Geometry.Create.Vector(pTops[i])).IRotate(pTops[i],Vector.ZAxis,property.OrientationAngle);
```

Comment:
Why do you only take one edge? Would also be good to check if the edges are actually assigned, and raise a warning about skipped check if not?

Reply:
We only need the outer edge (which is the first one) after some limited testing.

Reply:
I am afraid that would only be valid for round piles:
![image](https://github.com/BHoM/BHoM_Engine/assets/26874773/48f01bd1-fbf4-413e-bc49-291c65af45ea)

Reply:
Good point, I have added a `IJoin()` and then sorted by length to get the outer curve.